kernel/src/drivers/ps2/mod.rs

```diff
@@ -90,167 +93,159 @@ impl Ps2Controller {
+        Ps2Device {
+            state: DeviceState::Unavailable,
+            port,
+        }
+    }
+
     /// Tests this device to see if it is available
-    pub fn test(&mut self) -> bool {
-        let available = command_ret(if self.get_flag(DEVICE_SECOND_FLAG) {
+    pub fn test(&mut self) -> Result<bool, Ps2Error> {
+        let available = command_ret(if self.port == DevicePort::Mouse {
             ControllerReturnCommand::TestPort2
         } else {
             ControllerReturnCommand::TestPort1
-        }) == Some(0x0);
-        self.set_flag(DEVICE_AVAILABLE_FLAG, available);
-        available
+        })? == 0x0;
+
+        if available {
+            self.state = DeviceState::Available;
+        } else {
+            self.state = DeviceState::Unavailable;
+        }
+
+        Ok(available)
     }
 
     /// Enables this device
-    pub fn enable(&mut self) {
-        if self.get_flag(DEVICE_AVAILABLE_FLAG) {
-            command(if self.get_flag(DEVICE_SECOND_FLAG) {
+    pub fn enable(&mut self) -> Result<(), Ps2Error> {
+        if self.state == DeviceState::Available {
+            command(if self.port == DevicePort::Mouse {
                 ControllerCommand::EnablePort2
             } else {
                 ControllerCommand::EnablePort1
-            });
-            self.set_flag(DEVICE_ENABLED_FLAG, true);
+            })?;
+            self.state = DeviceState::Enabled;
         }
+        Ok(())
     }
 
     /// Disables this device
-    pub fn disable(&mut self) {
-        command(if self.get_flag(DEVICE_SECOND_FLAG) {
+    pub fn disable(&mut self) -> Result<(), Ps2Error> {
+        command(if self.port == DevicePort::Mouse {
             ControllerCommand::DisablePort2
         } else {
             ControllerCommand::DisablePort1
-        });
-        self.set_flag(DEVICE_ENABLED_FLAG, false);
+        })?;
+        self.state = DeviceState::Available;
+        Ok(())
     }
 
     /// Resets this device
-    pub fn reset(&mut self) {
-        self.command(DeviceCommand::Reset);
+    pub fn reset(&mut self) -> Result<(), Ps2Error> {
+        self.command(DeviceCommand::Reset)?;
+
+        Ok(())
     }
 
     /// Sends a command for this PS2 device and returns result
     #[allow(dead_code)] // Used by drivers depending on PS/2
-    pub fn command(&mut self, cmd: DeviceCommand) -> Option<u8> {
+    pub fn command(&mut self, cmd: DeviceCommand) -> Result<u8, Ps2Error> {
         // If second PS2 port, send context switch command
-        if self.get_flag(DEVICE_SECOND_FLAG) {
-            command(ControllerCommand::WriteInputPort2);
+        if self.port == DevicePort::Mouse {
+            command(ControllerCommand::WriteInputPort2)?;
         }
-        let mut result = Some(RESEND);
-        for _i in 0..4 {
-            if result != Some(RESEND) && result != None {
-                break;
+        for _ in 0..4 {
+            write_data(cmd as u8)?;
+            let result = read_data()?;
```

Comment:
Does this have to be in a separate variable?

Reply:
Nope, likely comes from legacy code.